### PR TITLE
Eliminate use of external commands from ocamltest

### PR DESCRIPTION
**Pull request by dra27**
The premise of this PR is work to speed up and make more correct the testsuite on Windows.

ocamltest intentionally has separate C stubs for process control. That probably should be the sole exception, however. The Windows symlink tests from fab457a8 introduced some duplication of the code for `unix_has_symlink`. The legitimate desire to avoid `ocamltest` depending on `Unix` causes `ocamltest` to make some slightly dodgy external calls:
- `mkdir` is used, which is just about portable, but slow on Windows (calling out to the shell always is)
- `ln -sf` is used, with Windows crippled never to use symlinks
- `rm -rf` is used, which is not portable.

There are some subtle impacts as well:
- Calling out to cmd so regularly can result in zombie processes in Cygwin if you press CTRL+C during the testsuite which locks the `_ocamltest` directory preventing its easy deletion. This becomes very noticeable on my box with 128 processes running in parallel...
- ocamltest should avoid needing to use `Filename.quote_command` because it's ~scary~ a complicated piece of code which should be being tested, not be part of the tester!

This PR adds the following primitives:
 - `caml_sys_mkdir` and `caml_sys_rmdir` which are exposed as `Sys.rmdir` and `Sys.mkdir` (the only argument for not exposing them I could think of was possibly the lack of `umask`, but we already have that situation for `open_out_gen`)
- `caml_sys_chmod` which is not exposed, on the basis that its inclusion makes the absence of `stat` feel strange (I dislike having the set function without the corresponding get)
- `caml_sys_has_symlink` and `caml_sys_symlink` neither of which are exposed in the standard library, for obvious reasons.

The Unix library is updated to use the new machinery where appropriate, without any (intended) change in semantics.

cc @shindere

### CODE REVIEW

**Comment by dra27** [comment]
I just rebased my parallelisation patches on to this PR - I haven't double-checked the old values, but this PR appears to take the parallel testsuite runtime down from just over 5 minutes to ~3:10 on my Windows system (i.e. just eliminating all those shell calls, and possibly reducing the amount of copying which is going on)

**Comment by nojb** [comment]
This is an excellent idea and I will review it shortly. I have a general question:

If I understand correctly, this PR moves a few functions from `Unix` to `Sys` so that they can be used in `ocamltest` (some of them are exposed in `Sys`, other are left unexposed). This also removes all non-process control related stubs from `ocamltest`.

This seems like a perfectly good idea (and `Sys.rmdir` and `Sys.mkdir` look useful). But the other possibility would be to depend directly on `Unix` and be done with it. Yes, we would extend slightly the "trusted surface" of `ocamltest`, but does it really matter in practice?

File: ocamltest/main.ml [hunk 1]
```diff
@@ -152,9 +152,10 @@ let test_file test_filename =
   let test_build_directory_prefix =
     get_test_build_directory_prefix test_directory in
   let clean_test_build_directory () =
-    ignore
-      (Sys.command
-         (Filename.quote_command "rm" ["-rf"; test_build_directory_prefix]))
+    try
+      Sys.rm_rf test_build_directory_prefix
+    with Sys.Break as e -> raise e
```
Comment (nojb) [inline ocamltest/main.ml:157]: Is this special handling of the `Break` exception due to the potentially lengthy nature of this operation?

**Review by nojb — COMMENTED** [review]
I did a first pass. Unless I missed something, the following functions from `Unix` are now duplicated in `Sys`:
- `chmod` 
- `mkdir`
- `rmdir`

As far as I can see the only difference between the two implementations is the exception raised in case of error (`Sys_error` for the functions in `Sys`, `Unix_error` for those in `Unix`). Could they be merged somehow? Does it matter?

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2020/07/21 13:07 -0700):
> This is an excellent idea and I will review it shortly. I have a general question:
>
> If I understand correctly, this PR moves a few functions from `Unix` to `Sys` so that they can be used in `ocamltest` (some of them are exposed in `Sys`, other are left unexposed). This also removes all non-process control related stubs from `ocamltest`.
>
> This seems like a perfectly good idea (and `Sys.rmdir` and `Sys.mkdir`
> look useful). But the other possibility would be to depend directly on
> `Unix` and be done with it. Yes, we would extend slightly the "trusted
> surface" of `ocamltest`, but does it really matter in practice?

Somehow yes, I think it does. Also, I think @xavierleroy agreed to add
'mkdiir' to Sys, that's something I had on my TODOLIST for a while so
it's really great that this is being implemented now and with such
benefices. Will review tomorrow but feel very happy about the PR in its
principles!

**Comment by xavierleroy** [comment]
I have mixed feelings about this proposal.

The reason why ocamltest doesn't depend on the Unix library was to reduce the dependencies between the device under test (OCaml and its core libraries) and the testing tool.

Adding a few Unix functions to Sys and using them in ocamltest is no better, dependency-wise, than using those functions from Unix.  It is slightly worse test-wise because the Unix functions have already been well tested and the new Sys functions have not.

So I'd rather have a discussion about which Unix functions are OK to use in ocamltest as an exception to our general policy.

Independently of this discussion and of ocamltest, we can have another discussion on adding `Sys.mkdir` and `Sys.rmdir`.  I don't think `chmod` belongs in `Sys`.

File: ocamltest/main.ml (hunk 1 above)
Comment (dra27) [reply ocamltest/main.ml:157]: Well, in a command-line tool, you never want to lose `Sys.Break` in internal handlers, yes. Imagine this is being done an NFS link or something which is timing out, for example.

**Comment by shindere** [comment]
Xavier Leroy (2020/07/22 00:10 -0700):
> I have mixed feelings about this proposal.
>
> The reason why ocamltest doesn't depend on the Unix library was to reduce the dependencies between the device under test (OCaml and its core libraries) and the testing tool.
>
> Adding a few Unix functions to Sys and using them in ocamltest is no better, dependency-wise, than using those functions from Unix.  It is slightly worse test-wise because the Unix functions have already been well tested and the new Sys functions have not.
>
> So I'd rather have a discussion about which Unix functions are OK to use in ocamltest as an exception to our general policy.
>
> Independently of this discussion and of ocamltest, we can have another
> discussion on adding `Sys.mkdir` and `Sys.rmdir`.  I don't think
> `chmod` belongs in `Sys`.

I seem to remember that, in the past, you (@xavierleroy) told me it
would be okay to add `mkdir` to Sys. Since then that's waiting on my
TODOLIST.

**Comment by dra27** [comment]
@nojb - The duplication of `rmdir`, `chmod` and `mkdir` I wondered about, but I remained "happy" with it on the basis that `Unix.chdir` and `Sys.chdir` have always been the same. Equally, they are simple functions and - based on the present code - highly unlikely to be altered.

@xavierleroy - the only functions added to the _Standard Library_ `Sys` module are `rmdir` and `mkdir` - `Sys.chmod` in ocamltest comes from `Ocamltest_stdlib.Sys`.

I completely agree with the premise of not using Unix in ocamltest. I think the test for moving things from Unix to the runtime should be that we eschew `Sys.command` in ocamltest (I remove `Ocamltest_stdlib.Sys.run_system_command` in this PR). I make an exception for running the `diff` command - it happens once and only on error. So, in the future if ocamltest needs to do something involving a shell command then we decide if it's really needed and, if so, do what's necessary _in the runtime_ to avoid that shell command.

The primitives fall into three categories:

1. The trivial - granted, I may have made a typo in the 44 lines of C added to `runtime/sys.c` for `chmod`, `rmdir` and `mkdir`, but the functions are almost identical to `caml_sys_chdir` already above them. The alternative is to fix the use of `rm -rf` to be `rd /s/q` on Windows. However, on my Windows system, creating and removing 3000 directories (that's what `ocamltest` does) takes 0.3 seconds with system calls and 11 minutes with `cmd /c` calls (it would still be 72 seconds even if we correctly used `cmd /c/d` instead).
2. The code-duplicating - `unix_has_symlink` was already duplicated in ocamltest, and this was already the cause of pain (see #9593). The alternative is to call a shell script to detect this instead.
3. The slightly more involved - `caml_sys_symlink` does not duplicate code between the runtime and Unix. The alternative is _always_ to copy or convert the Windows call to `mklink` (with the same caveat of being slow).

I reiterate - and with total respect to your implementation - that every time we use `Sys.command` we trust that `Filename.quote_command` is correct, while doubting whether trivial C functions are correct.

**Comment by nojb** [comment]
> So I'd rather have a discussion about which Unix functions are OK to use in ocamltest as an exception to our general policy.

@xavierleroy Would it be OK to use

- `Unix.rmdir`
- `Unix.chmod`
- `Unix.mkdir`
- `Unix.has_symlink` (stub already duplicated in `ocamltest`)
- `Unix.symlink`

in `ocamltest`? That would be enough for this PR.

**Comment by dra27** [comment]
Indeed, I misunderstood what you meant as adding discussion as to which primitives to move, not using Unix directly, but only allowing certain functions. I don't mind, the only caveat is that it is possible to configure the tree without the Unix library and it is possible to run the testsuite without the Unix library.

To further aid review, here's the diff of each new primitive against its `otherlibs/unix/function.c` counterpart:

```diff
-CAMLprim value caml_sys_mkdir(value path, value perm)
+CAMLprim value unix_mkdir(value path, value perm)
 {
   CAMLparam2(path, perm);
   char_os * p;
   int ret;
-  caml_sys_check_path(path);
+  caml_unix_check_path(path, "mkdir");
   p = caml_stat_strdup_to_os(String_val(path));
   caml_enter_blocking_section();
   ret = mkdir_os(p, Int_val(perm));
   caml_leave_blocking_section();
   caml_stat_free(p);
-  if (ret == -1) caml_sys_error(path);
+  if (ret == -1) uerror("mkdir", path);
   CAMLreturn(Val_unit);
 }
```
```diff
-CAMLprim value caml_sys_rmdir(value path)
+CAMLprim value unix_rmdir(value path)
 {
   CAMLparam1(path);
   char_os * p;
   int ret;
-  caml_sys_check_path(path);
+  caml_unix_check_path(path, "rmdir");
   p = caml_stat_strdup_to_os(String_val(path));
   caml_enter_blocking_section();
   ret = rmdir_os(p);
   caml_leave_blocking_section();
   caml_stat_free(p);
-  if (ret == -1) caml_sys_error(path);
+  if (ret == -1) uerror("rmdir", path);
   CAMLreturn(Val_unit);
 }
```
```diff
-CAMLprim value caml_sys_chmod(value path, value perm)
+CAMLprim value unix_chmod(value path, value perm)
 {
   CAMLparam2(path, perm);
   char_os * p;
   int ret;
-  caml_sys_check_path(path);
+  caml_unix_check_path(path, "chmod");
   p = caml_stat_strdup_to_os(String_val(path));
   caml_enter_blocking_section();
   ret = chmod_os(p, Int_val(perm));
   caml_leave_blocking_section();
   caml_stat_free(p);
-  if (ret == -1) caml_sys_error(path);
+  if (ret == -1) uerror("chmod", path);
   CAMLreturn(Val_unit);
 }
```
```diff
+#ifdef HAS_SYMLINK

-CAMLprim value caml_sys_symlink(value to_dir, value path1, value path2)
+CAMLprim value unix_symlink(value to_dir, value path1, value path2)
 {
-#if !defined(HAS_SYMLINK) && !defined(_WIN32)
-  caml_invalid_argument("symlink not implemented");
-#else
   CAMLparam3(to_dir, path1, path2);
-  char_os * p1;
-  char_os * p2;
+  char * p1;
+  char * p2;
   int ret;
-  caml_sys_check_path(path1);
-  caml_sys_check_path(path2);
-#ifdef _WIN32
-  if (!caml_win32_init_symlink())
-    caml_invalid_argument("symlink not available");
-#endif
-  p1 = caml_stat_strdup_to_os(String_val(path1));
-  p2 = caml_stat_strdup_to_os(String_val(path2));
+  caml_unix_check_path(path1, "symlink");
+  caml_unix_check_path(path2, "symlink");
+  p1 = caml_stat_strdup(String_val(path1));
+  p2 = caml_stat_strdup(String_val(path2));
   caml_enter_blocking_section();
-#ifdef _WIN32
-  ret = caml_win32_symlink(Bool_val(to_dir), p1, p2);
-#else
   ret = symlink(p1, p2);
-#endif
   caml_leave_blocking_section();
   caml_stat_free(p1);
   caml_stat_free(p2);
   if (ret == -1)
-    caml_sys_error(path2);
+    uerror("symlink", path2);
   CAMLreturn(Val_unit);
-#endif
 }
+
+#else
+
+CAMLprim value unix_symlink(value to_dir, value path1, value path2)
+{ caml_invalid_argument("symlink not implemented"); }
+
+#endif
```

**Comment by xavierleroy** [comment]
@dra27: sorry for the misunderstanding, I think we're now on the same page.

@nojb: nice list, thanks.  `mkdir`, `rmdir`, `chmod` are simple wrappers around standard library functions, so that's fine.  `symlink` is a little bit complicated under Windows (with all due respect to @dra27's heroic work), but `Unix.has_symlink` is generally false under Windows (as far as I know), so that should work well.

**Comment by dra27** [comment]
For CI, `Unix.has_symlink` should normally be true (servers run elevated by default) - AppVeyor runs the win-symlink tests, for example - and since #9593, most Windows 10 users will have it by default. It is indeed complicated...

I will do a revised PR for this shortly - amusingly, while poking around in `runtime/sys.c`, I have found several bugs in `Sys.is_directory` and `Sys.file_exists`...

**Comment by xavierleroy** [comment]
If we want to play it safe, we could use `Unix.symlink` under Unix and always copy files under Windows, at some performance cost, I know.

**Comment by gasche** [comment]
If I understand correctly, "the tests are unusable on my setup because symlink is broken" has never happened to anyone yet, and "I wish the testsuite was faster" is a common thought among the handful of unfortunate Windows-based developers. Are we optimizing for the right thing here? It would seems nicer to do the fast thing by default, with a fallback option --disable-symlinks for people suspecting that they are the lucky first winner to observe a symlink-related issue.

**Comment by dra27** [comment]
I would prefer the attitude "we can't make this work reliably on all platforms, so we'll do it **nowhere**" mainly because that encourages people to sort out the platform they care least about...! 🙂

That said, as with all these functions here, they're used extensively (i.e. in every single test) - if something's gone wrong in the Win32 implementation of `symlink`, you'll notice the entire testsuite exploding, so it at least won't be subtle and it's hard to conceive a failure mode of the function where the testsuite magically works in spite of it. The fix - at that time - would be a tiny patch in `actions_helpers.ml` to temporarily disable it while working out what's gone wrong.

**Comment by nojb** [comment]
Closing to focus on #9797, which is an improvement over this one.

**Pull request closed without merging**